fdbserver/MoveKeys.actor.cpp

```diff
@@ -1302,15 +1343,47 @@ ACTOR Future<Void> removeKeysFromFailedServer(Database cx,
 						continue;
 					}
 
+					TraceEvent("FailedServerRemoveBegin", serverID)
+					    .detail("Key", it.key)
+					    .detail("ValueSrc", describe(src))
+					    .detail("ValueDest", describe(dest));
+
 					// Update the vectors to remove failed server then set the value again
 					// Dest is usually empty, but keep this in case there is parallel data movement
 					src.erase(std::remove(src.begin(), src.end(), serverID), src.end());
 					dest.erase(std::remove(dest.begin(), dest.end(), serverID), dest.end());
-					TraceEvent(SevDebug, "FailedServerSetKey", serverID)
-					    .detail("Key", it.key)
-					    .detail("ValueSrc", describe(src))
-					    .detail("ValueDest", describe(dest));
-					tr.set(keyServersKey(it.key), keyServersValue(UIDtoTagMap, src, dest));
+					// If the last src server is to be removed, first check if there are dest servers who is
+					// hosting a read-write copy of the data, and move such dest servers to the src list.
+					if (src.empty() && !dest.empty()) {
+						std::vector<UID> newSources =
+						    wait(pickReadWriteServers(&tr, dest, KeyRangeRef(it.key, keyServers[i + 1].key)));
+						for (const UID& id : newSources) {
+							TraceEvent(SevWarn, "FailedServerAdditionalSourceServer", serverID)
+							    .detail("Key", it.key)
+							    .detail("NewSourceServerFromDest", id);
+							dest.erase(std::remove(dest.begin(), dest.end(), id), dest.end());
+							src.push_back(id);
+						}
+					}
+					// Remove the shard from keyServers/ if the src list is empty, and also remove the shard from all
+					// dest servers.
+					if (src.empty()) {
+						TraceEvent(SevWarn, "FailedServerRemoveRange", serverID)
+						    .detail("Key", it.key)
+						    .detail("ValueDest", describe(dest));
+						serversToRemoveRange.insert(serversToRemoveRange.end(), dest.begin(), dest.end());
```

Comment:
`serversToRemoveRange` seems not used anywhere?

Reply:
Removed, sorry, it was in the draft...